fem/qfunction.cpp

```diff
@@ -240,4 +242,45 @@ void QuadratureFunction::SaveVTU(const std::string &filename, VTKFormat format,
    SaveVTU(f, format, compression_level, field_name);
 }
 
+static double ReduceDouble(const Mesh *mesh, double value)
+{
+#ifdef MFEM_USE_MPI
+   if (auto *pmesh = dynamic_cast<const ParMesh*>(mesh))
+   {
+      MPI_Comm comm = pmesh->GetComm();
+      MPI_Allreduce(MPI_IN_PLACE, &value, 1, MPI_DOUBLE, MPI_SUM, comm);
+   }
+#endif
+   return value;
+}
+
+double QuadratureFunction::Integrate() const
+{
+   MFEM_VERIFY(vdim == 1, "Only scalar functions are supported.")
+   const double local_integral = (*this)*qspace->GetWeights();
+   return ReduceDouble(qspace->GetMesh(), local_integral);
+}
+
+void QuadratureFunction::Integrate(Vector &integrals) const
+{
+   integrals.SetSize(vdim);
+
+   const Vector &weights = qspace->GetWeights();
+   QuadratureFunction component(qspace);
+   const int N = component.Size();
+   const int VDIM = vdim; // avoid capturing 'this' in lambda body
+   const double *d_v = Read();
+
+   for (int vd = 0; vd < vdim; ++vd)
+   {
+      // Extract the component 'vd' into component.
+      double *d_c = component.Write();
+      mfem::forall(N, [=] MFEM_HOST_DEVICE (int i)
+      {
+         d_c[i] = d_v[vd + i*VDIM];
+      });
+      integrals[vd] = ReduceDouble(qspace->GetMesh(), component*weights);
+   }
+}
```

Comment:
This I believe is more of an issue with MFEM lacking a way to easily calculate reductions more than anything, but the above really shouldn't require constructing a whole new QF for this reduction. I would even argue that the even the `QuadratureSpaceBase::weights` variable is also not required here but that's only the case if there was a nice reduction operator kinda like RAJA's `RAJA::ReduceSum<ExecType, double>` type (although their current implementation has limitations...). Since, you can calculate everything required on the fly...

Reply:
Yeah, as far as I know, MFEM only really supports GPU reductions for dot products or sums. In this case, it could be modeled as a short and wide matrix times a vector, or as a kind of strided reduction… are there good ways of doing those kinds of reductions without a portability layer like RAJA? Would something like `cublasDgemv` be efficient? Or even `cublasDdot` takes a stride that could be used to eliminate the temporary. It would just be a little more boilerplate to handle HIP compatibility, etc.

Reply:
If we added some iterators to `QuadratureFunction` then you could use `std::accumulate` or the CUDA Thrust / rocThrust equivalent `thrust::reduce` operators to do this sorta thing.  However, I think moving to something like that would probably be a bit overkill for this sorta thing... I think with the #4065 merge then the temporary object creation is much less of an issue. I wouldn't worry about trying to fix this as the above though as there likely won't be a none abstraction layer equivalent available for all our platforms for a bit. 

As a side note, I did hear at SC23 that this will get better as apparently in C++26(?) there's supposedly going to be some better parallel execution aspects added in to the stdlib that's targeted more towards GPU vendors / some later patch changes into LLVM which will supposedly allow vendors to start allowing certain functions be host/device compatible.